## Preprocessing

In [6]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import tensorflow as tf

#  Import and read the charity_data.csv.
import pandas as pd
application_df = pd.read_csv("https://static.bc-edx.com/data/dl-1-2/m21/lms/starter/charity_data.csv")
application_df.head()

EIN                                      NAME APPLICATION_TYPE  \
0  10520599              BLUE KNIGHTS MOTORCYCLE CLUB              T10   
1  10531628    AMERICAN CHESAPEAKE CLUB CHARITABLE TR               T3   
2  10547893        ST CLOUD PROFESSIONAL FIREFIGHTERS               T5   
3  10553066            SOUTHSIDE ATHLETIC ASSOCIATION               T3   
4  10556103  GENETIC RESEARCH INSTITUTE OF THE DESERT               T3   

        AFFILIATION CLASSIFICATION      USE_CASE  ORGANIZATION  STATUS  \
0       Independent          C1000    ProductDev   Association       1   
1       Independent          C2000  Preservation  Co-operative       1   
2  CompanySponsored          C3000    ProductDev   Association       1   
3  CompanySponsored          C2000  Preservation         Trust       1   
4       Independent          C1000     Heathcare         Trust       1   

      INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  IS_SUCCESSFUL  
0              0                      N     5000              1  
1         1-9999                      N   108590              1  
2              0                      N     5000              0  
3    10000-24999                      N     6692              1  
4  100000-499999                      N   142590              1

## ***Optimization #3: Dropped additional columns of "SPECIAL_CONSIDERATIONS" and "ASK_AMT"***

In [7]:
# Drop the non-beneficial ID columns, 'EIN' and 'NAME'.
target = "IS_SUCCESSFUL"
application_df = application_df.drop(columns=["EIN", "NAME", "SPECIAL_CONSIDERATIONS", "ASK_AMT"])
application_df

APPLICATION_TYPE       AFFILIATION CLASSIFICATION      USE_CASE  \
0                  T10       Independent          C1000    ProductDev   
1                   T3       Independent          C2000  Preservation   
2                   T5  CompanySponsored          C3000    ProductDev   
3                   T3  CompanySponsored          C2000  Preservation   
4                   T3       Independent          C1000     Heathcare   
...                ...               ...            ...           ...   
34294               T4       Independent          C1000    ProductDev   
34295               T4  CompanySponsored          C3000    ProductDev   
34296               T3  CompanySponsored          C2000  Preservation   
34297               T5       Independent          C3000    ProductDev   
34298               T3       Independent          C1000  Preservation   

       ORGANIZATION  STATUS     INCOME_AMT  IS_SUCCESSFUL  
0       Association       1              0              1  
1      Co-operative       1         1-9999              1  
2       Association       1              0              0  
3             Trust       1    10000-24999              1  
4             Trust       1  100000-499999              1  
...             ...     ...            ...            ...  
34294   Association       1              0              0  
34295   Association       1              0              0  
34296   Association       1              0              0  
34297   Association       1              0              1  
34298  Co-operative       1          1M-5M              0  

[34299 rows x 8 columns]

In [8]:
# Determine the number of unique values in each column.
unique_values = application_df.nunique()
unique_values

APPLICATION_TYPE    17
AFFILIATION          6
CLASSIFICATION      71
USE_CASE             5
ORGANIZATION         4
STATUS               2
INCOME_AMT           9
IS_SUCCESSFUL        2
dtype: int64

In [9]:
# Look at APPLICATION_TYPE value counts for binning
print(application_df["APPLICATION_TYPE"].value_counts())

T3     27037
T4      1542
T6      1216
T5      1173
T19     1065
T8       737
T7       725
T10      528
T9       156
T13       66
T12       27
T2        16
T25        3
T14        3
T29        2
T15        2
T17        1
Name: APPLICATION_TYPE, dtype: int64


## ***Optimization #3: Kept the cutoff_value at 10***

In [10]:
# Choose a cutoff value and create a list of application types to be replaced
# use the variable name `application_types_to_replace`
cutoff_value = 10
application_types_to_replace = application_df["APPLICATION_TYPE"].value_counts()[application_df["APPLICATION_TYPE"].value_counts()
 < cutoff_value].index.tolist()

# Replace in dataframe
for app in application_types_to_replace:
    application_df['APPLICATION_TYPE'] = application_df['APPLICATION_TYPE'].replace(app,"Other")

# Check to make sure binning was successful
application_df['APPLICATION_TYPE'].value_counts()

T3       27037
T4        1542
T6        1216
T5        1173
T19       1065
T8         737
T7         725
T10        528
T9         156
T13         66
T12         27
T2          16
Other       11
Name: APPLICATION_TYPE, dtype: int64

In [11]:
# Look at CLASSIFICATION value counts for binning
classification_counts = application_df["CLASSIFICATION"].value_counts()
classification_counts

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
         ...  
C4120        1
C8210        1
C2561        1
C4500        1
C2150        1
Name: CLASSIFICATION, Length: 71, dtype: int64

In [12]:
# You may find it helpful to look at CLASSIFICATION value counts >1
# Filter and print value counts for CLASSIFICATION with counts > 1
classification_counts_gt1 = application_df["CLASSIFICATION"].value_counts()[application_df["CLASSIFICATION"].value_counts() > 1]
print(classification_counts_gt1)


C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
C7000      777
C1700      287
C4000      194
C5000      116
C1270      114
C2700      104
C2800       95
C7100       75
C1300       58
C1280       50
C1230       36
C1400       34
C7200       32
C2300       32
C1240       30
C8000       20
C7120       18
C1500       16
C1800       15
C6000       15
C1250       14
C8200       11
C1238       10
C1278       10
C1235        9
C1237        9
C7210        7
C2400        6
C1720        6
C4100        6
C1257        5
C1600        5
C1260        3
C2710        3
C0           3
C3200        2
C1234        2
C1246        2
C1267        2
C1256        2
Name: CLASSIFICATION, dtype: int64


## ***Optimization #3: Kept the cutoff_value at 1000***

In [15]:
# Choose a cutoff value and create a list of classifications to be replaced
# use the variable name `classifications_to_replace`
cutoff_value_classification = 1000
classifications_to_replace = application_df["CLASSIFICATION"].value_counts()[application_df["CLASSIFICATION"].value_counts() < cutoff_value_classification].index.tolist()

# Replace in dataframe
for cls in classifications_to_replace:
    application_df['CLASSIFICATION'] = application_df['CLASSIFICATION'].replace(cls,"Other")

# Check to make sure binning was successful
print(application_df['CLASSIFICATION'].value_counts())

C1000    17326
C2000     6074
C1200     4837
Other     2261
C3000     1918
C2100     1883
Name: CLASSIFICATION, dtype: int64


In [16]:
# Convert categorical data to numeric with `pd.get_dummies`
application_df_dummies = pd.get_dummies(application_df)
application_df_dummies.head()

STATUS  IS_SUCCESSFUL  APPLICATION_TYPE_Other  APPLICATION_TYPE_T10  \
0       1              1                       0                     1   
1       1              1                       0                     0   
2       1              0                       0                     0   
3       1              1                       0                     0   
4       1              1                       0                     0   

   APPLICATION_TYPE_T12  APPLICATION_TYPE_T13  APPLICATION_TYPE_T19  \
0                     0                     0                     0   
1                     0                     0                     0   
2                     0                     0                     0   
3                     0                     0                     0   
4                     0                     0                     0   

   APPLICATION_TYPE_T2  APPLICATION_TYPE_T3  APPLICATION_TYPE_T4  ...  \
0                    0                    0                    0  ...   
1                    0                    1                    0  ...   
2                    0                    0                    0  ...   
3                    0                    1                    0  ...   
4                    0                    1                    0  ...   

   ORGANIZATION_Trust  INCOME_AMT_0  INCOME_AMT_1-9999  \
0                   0             1                  0   
1                   0             0                  1   
2                   0             1                  0   
3                   1             0                  0   
4                   1             0                  0   

   INCOME_AMT_10000-24999  INCOME_AMT_100000-499999  INCOME_AMT_10M-50M  \
0                       0                         0                   0   
1                       0                         0                   0   
2                       0                         0                   0   
3                       1                         0                   0   
4                       0                         1                   0   

   INCOME_AMT_1M-5M  INCOME_AMT_25000-99999  INCOME_AMT_50M+  \
0                 0                       0                0   
1                 0                       0                0   
2                 0                       0                0   
3                 0                       0                0   
4                 0                       0                0   

   INCOME_AMT_5M-10M  
0                  0  
1                  0  
2                  0  
3                  0  
4                  0  

[5 rows x 45 columns]

In [17]:
# Split our preprocessed data into our features and target arrays
y = application_df_dummies["IS_SUCCESSFUL"].values
X = application_df_dummies.drop("IS_SUCCESSFUL", axis=1).values

print("Shape of X:", X.shape)
print("Shape of y:", y.shape)

# Split the preprocessed data into a training and testing dataset
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=78)

print("Shape of X_train:", X_train.shape)
print("Shape of X_test:", X_test.shape)
print("Shape of y_train:", y_train.shape)
print("Shape of y_test:", y_test.shape)

Shape of X: (34299, 44)
Shape of y: (34299,)
Shape of X_train: (25724, 44)
Shape of X_test: (8575, 44)
Shape of y_train: (25724,)
Shape of y_test: (8575,)


In [18]:
# Create a StandardScaler instances
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()

# Fit the StandardScaler
X_scaler = scaler.fit(X_train)

# Scale the data
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

## Compile, Train and Evaluate the Model

## ***Optimization #3: Changed the second layer activation from relu to sigmoid. Kept the model at 2 hidden layers. Changed units to 20 on first hidden layer.***

In [20]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
nn = tf.keras.models.Sequential()

# First hidden layer / removed: input_dim=len(X_train_numeric.columns)
nn.add(tf.keras.layers.Dense(units=20, activation='relu'))

# Second hidden layer - this time with activiation = sigmoid and not relu
nn.add(tf.keras.layers.Dense(units=5, activation='sigmoid'))

# Output layer
nn.add(tf.keras.layers.Dense(units=1, activation='sigmoid'))

# Build the model
nn.build(input_shape=(None, len(X_train[0])))

# Check the structure of the model
nn.summary()

Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_3 (Dense)             (None, 20)                900       
                                                                 
 dense_4 (Dense)             (None, 5)                 105       
                                                                 
 dense_5 (Dense)             (None, 1)                 6         
                                                                 
Total params: 1011 (3.95 KB)
Trainable params: 1011 (3.95 KB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


In [21]:
# Compile the model
nn.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

## ***Optimization #3: Changed the epochs from 50 to 100***

In [22]:
# Train the model / verbose=2
nn.fit(X_train_scaled, y_train, epochs=100)

Epoch 1/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5983 - accuracy: 0.6999
Epoch 2/100
804/804 [==============================] - 1s 2ms/step - loss: 0.5645 - accuracy: 0.7303
Epoch 3/100
804/804 [==============================] - 1s 2ms/step - loss: 0.5589 - accuracy: 0.7321
Epoch 4/100
804/804 [==============================] - 3s 4ms/step - loss: 0.5557 - accuracy: 0.7314
Epoch 5/100
804/804 [==============================] - 4s 5ms/step - loss: 0.5532 - accuracy: 0.7326
Epoch 6/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5521 - accuracy: 0.7331
Epoch 7/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5508 - accuracy: 0.7328
Epoch 8/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5499 - accuracy: 0.7333
Epoch 9/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5495 - accuracy: 0.7332
Epoch 10/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5490 - accura

In [23]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 0s - loss: 0.5525 - accuracy: 0.7254 - 479ms/epoch - 2ms/step
Loss: 0.5524521470069885, Accuracy: 0.7253644466400146


## ***Optimization #3:***
Starter code optimization: 268/268 - 2s - loss: 0.5521 - accuracy: 0.7271 - 2s/epoch - 8ms/step
Loss: 0.5521000027656555, <br>
Accuracy: 0.7271137237548828

Optimization #3: 268/268 - 0s - loss: 0.5525 - accuracy: 0.7254 - 479ms/epoch - 2ms/step
Loss: 0.5524521470069885, <br>
Accuracy: 0.7253644466400146


Comparison: Optmization #3 changes decreased the accuracy by .002

In [24]:
# Export our model to HDF5 file
# nn.save("AlphabetSoupCharity.h5")

from google.colab import drive

# Mount Google Drive
drive.mount('/content/gdrive')

# Save the model to Google Drive
nn.save('/content/gdrive/MyDrive/AlphabetSoupCharity_Optimization3.h5')
nn.save('/content/gdrive/MyDrive/AlphabetSoupCharity_Optimization3.keras')

Mounted at /content/gdrive


/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py:3079: UserWarning: You are saving your model as an HDF5 file via `model.save()`. This file format is considered legacy. We recommend using instead the native Keras format, e.g. `model.save('my_model.keras')`.
  saving_api.save_model(
